### ISSUE TITLE
Testing results

### ISSUE THREAD

**Issue by Rafique-tech** [mahmoudnafifi/C5#14]
Hi Mahmoud, 
For my work, one of my tasks is to reproduce the results mentioned in your paper. For this, I am testing the provided pre-trained model on the INETL-TAU dataset (7022 images) with m=7. Since the images are already black-level subtracted as mentioned on the dataset website, I am directly passing resized PNG images (384×256) to the model along with corresponding .json files (illuminant information). I am also using cross-validation but without the G multiplier for testing.  My obtained results are as: Mean: 2.61, Median: 1.77, Best25: 0.57, Worst25: 1.44, Worst05: 2.16, Tri: 1.95, and Max: 28.39.

There is slight variation in results except the Worst25 which has a lot. As per my understanding, one reason could be the random sample selection nature of cross-validation. Is it so? or is there any other important step, I am missing?

Another thing to be mentioned, during the test I didn't mask out the color checker present in the scenes that you mentioned in the paper. Could you please provide details on it, how you did that? Because I think for the masking the coordinates for the color checker in each scene should be known.

**Comment by mahmoudnafifi** [comment]
Hi, there was a mistake in the evaluation script provided in this repo -- it does not affect the reported results, as the script used for evaluation in the paper does not have this bug. I fixed this bug in the public evaluation code [here](https://github.com/mahmoudnafifi/C5/commit/dca92361d43c8e737697fb27ac6eb9aa5d640bc4). Please try again with the updated evaluation script, and let me know if you still encounter any issues with 'Worst25'.

Regarding the second point, yes, we masked out color checker pixels in datasets that include color charts in the scenes. In these datasets, the color chart coordinates are provided by the authors. In Intel-tau case, there are no color charts present in the images, so this step is unnecessary.

**Comment by RafiqueA03** [comment]
Hi,
Thanks for the update. Now the obtained results are almost the same.

However, there is another question. The provided pre-trained model is trained on data_num (m)=7. Does this mean that for testing the data_num should explicitly be equal to 7? Because if I put data_num less or greater than 7, I get the error "RuntimeError: Error(s) in loading state_dict for network:".

Can you please clarify? should the model be trained and tested on the same value of data_num? For example, in the training section of the paper for m=9, it is mentioned that, **"then randomly select eight additional input images for each query image from the training set for use as additional input images"**.

Also how encoder blocks are being generated. From visualization in the paper, it appears that for m images encoder blocks will be m.

**Comment by mahmoudnafifi** [comment]
Yes testing should be performed using the same m, which refers to the number of additional histograms randomly selected + the input histogram. m=7 means we use 6 additional histograms. The number of encoders are equal to m.

**Comment by RafiqueA03** [comment]
Dear Mahmoud,
Thanks for the previous response.

I want to train the model with m=1 (only a single image) on the NUS and Gheleri-Shi datasets without proposed data augmentation. To train the model the input can be either 16-bit or 8-bit but as per your response to one question here on GitHub, it is recommended to use 16-bit which should be black-level normalized. I have a few questions and it would be very kind of you to clear them so that I can start with model training.

**Pre-processing of input to model for training:**

1. For black level normalization of the NUS dataset both the black_level and saturation are provided for each camera hence for the normalization I am using: **(image-black_level)/(saturation-black_level)**, following your response to one of the questions, here on GitHub. With this since the pixel range changes to: 0 to 1 hence cv2.imwirte clips them to zero i.e. the resultant image is zero. To avoid this I am normalizing the black-level normalized image to 16-bit (cv2.normalize(normalized_image, dst=None, alpha=0, beta=65535,norm_type=cv2.NORM_MINMAX, dtype=cv2.CV_16U)). I want to ensure this approach is fine i.e. will it affect the training? or is there any other way you have utilized to avoid this nature of cv2.imwrite?

2. For the Gheleri-Shi dataset:
- RGB values for illuminant are given I think these need to be normalized using r+g+b for the json files. 
- Also, only black-level values are provided for both cameras (saturation is missing or at least couldn't be found) so in that case how has black_level normalization been applied? One way is to use the provided Matlab file in the dataset for removing the black_level (canon1d has 0 and canon5d has 129 black_level value). 
- Apart from that, this dataset has 12-bit images as described on the dataset website. So should images be normalized to 16-bit or 8-bit after black_level removal as the C5 expects an input of 16 or 8-bit.

**Model training on m=1 without data augmentation technique**

1. To start with this as per my understanding only the value of data_num is supposed to be kept to 1 for training the model keeping other parameters default as defined in the script training.py. Is there anything else I should consider? 
2. Is it possible to compare the results of such a model (trained on NUS and Gheleri-Shi dataset with m=1 without data augmentation) to be sure that the ones I will obtain are accurate so that further steps can be taken for my work?

Sorry for the too-detailed questions.
I will be very grateful to you for answering these.
